### PR TITLE
Run mypy with current Twisted

### PR DESCRIPTION
**Pull request by wsanchez**
Un-pin Twisted when running Mypy.

It's pinned right now because with Twisted 21, we have numerous new errors.

### CODE REVIEW

File: src/klein/_request_compat.py [hunk 1]
```diff
@@ -93,15 +92,10 @@ def bodyAsFount(self) -> IFount:
...
 
-    def bodyAsBytes(self) -> Deferred:
+    async def bodyAsBytes(self) -> bytes:
         if self._state.cachedBody is not None:
-            return succeed(self._state.cachedBody)
-
-        def cache(bodyBytes: bytes) -> bytes:
-            self._state.cachedBody = bodyBytes
-            return bodyBytes
+            return self._state.cachedBody
```
Comment (graingert) [inline src/klein/_request_compat.py:97]:
```suggestion
            return self._state.cachedBody  # pragma: no cover
```

File: src/klein/test/test_headers_compat.py [hunk 2]
```diff
@@ -21,7 +19,7 @@
 try:
     from twisted.web.http_headers import _sanitizeLinearWhitespace
 except ImportError:
```
Comment (graingert) [inline src/klein/test/test_headers_compat.py:21]:
```suggestion
except ImportError:  # pragma: no cover
```

File: src/klein/test/test_headers_compat.py [hunk 3]
```diff
@@ -30,12 +28,9 @@ def _twistedHeaderNormalize(value: str) -> str:
     Twisted version.
     """
     if _sanitizeLinearWhitespace is None:
```
Comment (graingert) [inline src/klein/test/test_headers_compat.py:30]:
```suggestion
    if _sanitizeLinearWhitespace is None:  # pragma: no cover
```

File: src/klein/test/test_resource.py [hunk 4]
```diff
@@ -20,118 +21,126 @@
...
         request.finish()
         return succeed(None)
-    elif result is NOT_DONE_YET:
-        if request.finished or not notifyFinish:
-            return succeed(None)
-        else:
-            return request.notifyFinish()
+
+    if result is not NOT_DONE_YET:  # type: ignore[comparison-overlap]
+        raise AssertionError("unreachable")
```
Comment (graingert) [inline src/klein/test/test_resource.py:138]:
```suggestion
        raise AssertionError("unreachable")  # pragma: no cover
```

File: src/klein/_form.py [hunk 5]
```diff
@@ -151,20 +152,22 @@ def asTags(self) -> Iterable[Tag]:
...
-                    self.error.message
+                    self.error.message  # type: ignore[arg-type]
                 )
             )
         if self.formLabel:
             yield tags.label(self.formLabel, ": ", input_tag, *error_tags)
         else:
             yield input_tag
-            yield error_tags
+            yield error_tags  # type: ignore[misc]
```
Comment (graingert) [inline src/klein/_form.py:170]:
```suggestion
            yield from error_tags
```

File: src/klein/test/test_request_compat.py [hunk 6]
```diff
@@ -143,19 +147,19 @@ def test_bodyAsBytes(self, data: bytes) -> None:
         """
         legacyRequest = self.legacyRequest(body=data)
         request = HTTPRequestWrappingIRequest(request=legacyRequest)
-        body = self.successResultOf(request.bodyAsBytes())
+        body = self.successResultOf(ensureDeferred(request.bodyAsBytes()))
 
         self.assertEqual(body, data)
 
-    def test_bodyAsBytesCached(self) -> None:
+    async def test_bodyAsBytesCached(self) -> None:
```
Comment (graingert) [inline src/klein/test/test_request_compat.py:154]: @wsanchez async tests are not tested yet https://github.com/twisted/twisted/pull/1399

**Review by graingert — APPROVED** [review]

**Comment by graingert** [comment]
@wsanchez please review my additional changes before merging

File: src/klein/_form.py [hunk 7]
```diff
@@ -367,7 +372,7 @@ def lookupRenderMethod(self, name: str) -> NoReturn:
         """
         raise MissingRenderMethod(self, name)
 
-    def render(self, request: IRequest) -> Tag:
+    def render(self, request: object) -> Tag:
```
Comment (wsanchez) [inline src/klein/_form.py:375]: Why this change?

File: src/klein/_form.py [hunk 8]
```diff
@@ -393,9 +396,8 @@ def glue(self) -> Iterable[Tag]:
 
         @return: some HTML elements in the form of renderable objects for
             L{twisted.web.template}
-        @rtype: L{twisted.web.template.Tag}, or L{list} thereof.
         """
-        return self._fieldForCSRF().asTags()
+        return list(self._fieldForCSRF().asTags())
```
Comment (wsanchez) [inline src/klein/_form.py:400]: This suggests that the type hint in Twisted is too narrow?

Comment (wsanchez) [reply src/klein/_form.py:400]: OK, looking at the code, I think not. So now I'm wondering if copying to a new list is necessary here…

Comment (wsanchez) [reply src/klein/_form.py:400]: OK will worry about this later

File: src/klein/_form.py (hunk 7 above)
Comment (wsanchez) [reply src/klein/_form.py:373]: Mypy passes if I revert this, and it should be an `IRequest`…

File: src/klein/_iform.py [hunk 9]
```diff
@@ -3,7 +3,7 @@ class ValidationError(Exception):
     A L{ValidationError} is raised by L{Field.extractValue}.
     """
 
-    def __init__(self, message: str) -> None:
+    def __init__(self, message: object) -> None:
```
Comment (wsanchez) [inline src/klein/_iform.py:6]: Why this change? Mypy passes if reverted.

File: src/klein/test/test_request_compat.py (hunk 6 above)
Comment (wsanchez) [reply src/klein/test/test_request_compat.py:177]: Whoops, I knew that and forgot the decorator, thanks

File: src/klein/_iform.py (hunk 9 above)
Comment (wsanchez) [reply src/klein/_iform.py:6]: OK I see it
